### PR TITLE
Feature 1792 - users can choose to recieve password resets via SMS

### LINKED ISSUE #1792: users can choose to receive password reset via SMS
Part of epic #1017, dependent on #1785, #1790, #1791, #1780, and #1782

AS a **volunteer/supervisor/admin**,
I WANT the option to receive a password reset link via SMS,
SO THAT I am not restricted by receiving it only via email.

### How it works now
Right now, all users only have the option to receive a password reset link by email.

### How it should work instead
All users should have the option to receive a password reset link via SMS. 

### Acceptance criteria
- [ ] on the "forgot password" page, there should be a field for mobile number
- [ ] the field should only accept numbers
- [ ] field should only accept mobile numbers that are the appropriate length of a US mobile number (ten digits without +1 in the front, 11 digits if 1 is included as the first digit)
- [ ] the text "or" should separate the email and mobile number fields so it is obvious the user only needs to enter **one**
- [ ] the phone number entered must already be associated with an existing user account
- [ ] clicking "send me reset password instructions" after entering an existing mobile number triggers an SMS notification containing a reset password link
- [ ] that SMS should read something like:
_Click here to reset your password on casavolunteetracking.org [link]_
- [ ] clicking "send me password instructions" after entering a mobile number overrides previous opt-out preferences (so that the user can receive the one-time notification they are requesting)

### PR DESCRIPTION
**Pull request by xihai01**
### What github issue is this PR for, if any?
Resolves #1792 

### What changed, and why?
CASA users can now send password resets via SMS in addition to email.
Users can have the flexible options of getting a reset link via:
- Just email
- Just phone number
- Both email and phone number

To achieve this:
- A phone number field is added in the front end
- A passwords controller with a create action is implemented to override Devise's passwords controller. This is needed to add in the functionality of sending via SMS.

### How is this tested? (please write tests!) 💖💪
- A view spec is created to test front end aspect (i.e additional text and phone number field is present on page)
- A system spec is created to test user flow (forgot password -> enter email and/or phone number -> click send -> redirect with a notice) and rendering of any form validation errors
- A request spec is created to test our custom route that it calls short io and twilio services
 and redirects if no form errors occur

### Screenshots please :)
demo

#### SMS received via my phone (short link was visible thanks to @7riumph 's branded domain)
SMS

### Feelings gif (optional)
excitment
I really enjoyed this issue - it allowed me to dive deeper into the internals of Devise

### CODE REVIEW

File: spec/system/devise/passwords/new_spec.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+require "rails_helper"
+
+RSpec.describe "users/passwords/new", type: :system do
+  before do
+    visit root_path
+    click_on "Forgot your password?"
+  end
+
+  it "displays text above form fields" do
+    expect(page).to have_text("Please enter email or phone number to recieve reset instructions.")
```
Comment (xihai01) [inline spec/system/devise/passwords/new_spec.rb:10]: TO DO: move this test to view spec

**Comment by xihai01** [comment]
@compwron @harsohailB @7riumph rspec is failing because the db is missing a "user reminder times" table and the lib/tasks tests depend on it. Looking at the schema, it seems to me someone refactored the "users reminder times" to "user case contact types reminders". Any clue to what's going on here?
Screenshot from 2022-07-16 11-18-39
Screenshot from 2022-07-16 11-25-42

**Comment by xihai01** [comment]
I'll work on fixing the code climate issues. Please don't merge yet or until everything is green. :)

**Comment by harsohailB** [comment]
> it seems to me someone refactored the "users reminder times" to "user case contact types reminders"

@xihai01 It should be the opposite. I refactored "user case contact types reminders" table to "users reminder times" to make it re-usable for all reminder times. #3735

File: app/controllers/users/passwords_controller.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    if email.blank? && phone_number.blank?
+      resource.errors.add(:base, "Please enter at least one field.")
+    end
+
+    if !email.blank? && !User.find_by(email: email)
+      resource.errors.add(:base, "Email not found")
+    end
+
+    validation = valid_phone_number(phone_number)
+    if validation[0]
```
Comment (harsohailB) [inline app/controllers/users/passwords_controller.rb:20]: Can we do something like this instead so its more readable than indexing:
```ruby
phone_number_is_valid, error_message = valid_phone_number(phone_number)

if phone_number_is_valid
    error_message = "Phone number not found" if User.find_by(phone_number: phone_number)
end

if error_message
    resource.errors.add(:phone_number, error_message)
end
```

File: app/controllers/users/passwords_controller.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    end
+
+    # re-render and display any errors
+    if resource.errors.any?
+      respond_with(resource)
+      return
+    end
+
+    # otherwise, send reset email and sms
+    @resource = email.blank? ? User.find_by(phone_number: phone_number) : User.find_by(email: email)
```
Comment (harsohailB) [inline app/controllers/users/passwords_controller.rb:33]: Seems like you are already querying `User.find_by(phone_number: phone_number)` and `User.find_by(email: email)` above in the file to add error if the queries doesn't find anything. Maybe put these in a variable so you only call them once and reuse them if the query does find something?

File: app/controllers/users/passwords_controller.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+      sms_params = {
+        From: @resource.casa_org.twilio_phone_number,
+        Body: body_msg,
+        To: phone_number
+      }
+      twilio_service.send_sms(sms_params)
+    end
+
+    redirect_to after_sending_reset_password_instructions_path_for(resource_name), notice: "You will receive an email or SMS with instructions on how to reset your password in a few minutes."
```
Comment (harsohailB) [inline app/controllers/users/passwords_controller.rb:60]: Just a question: What happens if they put in both email and phone number? Do they get notified through email and sms?

File: app/controllers/users/passwords_controller.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  include ApplicationHelper
+  include PhoneNumberHelper
+  include SmsBodyHelper
+
+  def create
+    email = params[resource_name][:email]
+    phone_number = params[resource_name][:phone_number]
+    reset_token = nil
+
+    if email.blank? && phone_number.blank?
```
Comment (harsohailB) [inline app/controllers/users/passwords_controller.rb:11]: I think you should have 2 clear paths for email and sms based on what is filled out. Right now, its hard to follow the logic as we have a lot `if` statements that keep checking `email.blank?` and `phone_number.blank?`. Maybe:

```ruby
if !email.blank? && valid_email
   # run this code which  only has to do with the email
end

if !phone_number.blank? && valid_phone_number
   # run this code which  only has to do with the sms
end
```

**Review by harsohailB — COMMENTED** [review]
Really nice job! Glad you enjoyed this because I was having a really hard time wrapping my head around `Devise`. Just a few comments about code separation and reusability in the passwords controller.

File: app/controllers/users/passwords_controller.rb (hunk 4 above)
Comment (xihai01) [reply app/controllers/users/passwords_controller.rb:60]: yes, that is intended. If user supplies both email and phone number, they get both an email and SMS

**Comment by xihai01** [comment]
TO DO STILL: 
- create action is apparently still too complex for a human to understand -> need refactoring
- migrate request spec to a controller spec to pass that spec checker :hankey: 
- fix error with failing lib/tasks tests

**Comment by compwron** [comment]
This is exciting!
CI is failing on lint and spec

**Review by compwron — APPROVED** [review]

File: app/controllers/users/passwords_controller.rb [hunk 6]
```diff
@@ -0,0 +1,71 @@
+class Users::PasswordsController < Devise::PasswordsController
```
Comment (compwron) [inline app/controllers/users/passwords_controller.rb:1]: eek

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -44,4 +44,16 @@ def flash_class(level)
   def og_tag(type, options = {})
     tag.meta(property: "og:#{type}", **options)
   end
+
+  def resource_name
```
Comment (compwron) [inline app/helpers/application_helper.rb:48]: for devise ?

**Review by compwron — APPROVED** [review]

**Review by 7riumph — COMMENTED** [review]
🔥🔥🔥

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (xihai01) [reply app/helpers/application_helper.rb:48]: yes, for devise. The devise view calls some internal methods like resource and resource_name. So we need to map these to use them in our controllers and specs
